lib/plugins/chat.js

```diff
@@ -108,43 +108,52 @@ function inject (bot, options) {
 
   addDefaultPatterns()
 
-  // Pre 1.19
-  bot._client.on('chat', (packet) => {
-    const msg = ChatMessage.fromNotch(packet.message)
-    const chatPositions = {
-      0: 'chat',
-      1: 'system',
-      2: 'game_info'
-    }
-    bot.emit('message', msg, chatPositions[packet.position], packet.sender, /* verified */ null)
-    bot.emit('messagestr', msg.toString(), chatPositions[packet.position], msg, packet.sender, /* verified */ null)
-    // Position 2 is the action bar
-    if (packet.position === 2) bot.emit('actionBar', msg, null)
-  })
+  bot._client.on('playerChat', (data) => {
+    const message = data.formattedMessage
+    const verified = data.verified
+    let msg
+    if (bot.supportFeature('clientsideChatFormatting')) {
+      let content
 
-  // 1.19+
-  bot._client.on('player_chat', (packet) => {
-    const message = packet.unsignedChatContent || packet.signedChatContent
-    let verified = false
-    const sender = bot.uuidToUsername[packet.senderUuid]
-    if (sender) {
-      const { profileKeys } = bot.players[sender]
-      if (profileKeys) verified = bot._client.verifyMessage(profileKeys.publicKey, packet)
-    }
-    const parameters = {
-      sender: JSON.parse(packet.senderName),
-      content: JSON.parse(message)
+      if (message) content = JSON.parse(message)
+      else if (options.allowInsecureChat && data.unsignedContent) content = JSON.parse(data.unsignedContent)
+      else content = { text: data.plainMessage }
```

Comment:
Instead of this, you can emit both the message and the unsigned content for users that want to get access to the data. When we implemented this initially the idea was that only one would be relevant -- but if both are valid then the users can handle which message to look at. To make it not breaking you can add a .unsignedContent to the emitted ChatMessage

Reply:
If we're just attempting to mirror the vanilla client then only one will be valid.

Reply:
Yes, one will be the default but the other still available from the data. Only should be unexposed if it's one or the other (ie this or that)

Reply:
I have (more or less) done what you requested, but I honestly have to disagree. For NMP, sure, we should expose everything, but for Mineflayer we should really only expose what the Vanilla client would expose.

Reply:
mineflayer is mainly to serve as a higher level abstraction from the lower level networking things, yes. But that doesn't mean valuable information shouldn't be exposed or put behind flags. If there's an stable API to get at the data, then in this case think it would make sense as an [underscore method](https://github.com/PrismarineJS/mineflayer/blob/master/docs/unstable_api.md). For features like this, it can be left to the user libraries to do what they need. 

For example, in prismarine-web-client we might want to show the unsigned content when a user clicks on a message for additional context. Shouldn't be hard to do or require special config on the bot / listening to raw packets. The previous `message: signed || unsigned` was put under the assumption one would be empty and one not, so user code didn't need to worry about it. As you mention in the nmp PR, both these fields are set messages, but one may be server formatted and the other the players' original message. 

For simplicity/compatibility we want to put out the message that would appear to a vanilla client. But we can also send out the original message in a documented field in the chat object (or emit it as a new parameter).

Reply:
I guess if anyone can find a usecase for it as you have here, then let's implement it. I would just personally refer to things that aren't accessible in the Vanilla client as "low-level" and Mineflayer is not a low-level library.

Gonna push a new commit to implement the changes but I do not think it makes sense.

Reply:
Sure, we can just not expose this. My main point of objection is there should not be a special flag that changes the presence/output of a event. There should just be a new field to the object/event. 

The flags we added to nmp are required because they impact login procedure and can't be changed later on. For example, disableChatSigning on the client skips a call to pauth for dealing with certs. enforceSecureProfile on the server affects the login procedure for clients entering.

Reply:
By doing it that way we leave chat formatting to higher-level libraries which I think is not optimal. I can get behind the way it is implemented currently, but if we decide to just emit multiple messages and let libraries choose which to select adds unnecessary complication.

I vote we keep it this way. It's the best of both worlds

Reply:
Not a new event, in the above reply I meant a new field or parameter to the ChatMessage object we're emitting.

Reply:
Alright. Then we're on the same page. That is how it is now